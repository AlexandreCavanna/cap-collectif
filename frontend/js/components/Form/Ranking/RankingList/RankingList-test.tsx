/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import RankingList from './RankingList'

const baseRankingList = {
  id: '123-rakintlist',
  dataForm: {
    id: '12',
    helpText: "Texte d'aide",
    required: false,
    choices: [
      {
        id: '24',
        label: 'Maxime Arrouard',
      },
      {
        id: '25',
        label: 'Spylou Super Sayen',
      },
      {
        id: '26',
        label: 'Cyril Lage',
      },
      {
        id: '27',
        label: 'Jpec Superman',
      },
    ],
    values: null,
  },
  isDisabled: false,
  onChange: jest.fn(),
}
const rankingList = {
  basic: { ...baseRankingList },
  isDisabled: { ...baseRankingList, isDisabled: true },
}
describe('<RankingList />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<RankingList {...rankingList.basic} hasIllustrationDisplayed />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render correctly when disabled', () => {
    const wrapper = shallow(<RankingList {...rankingList.isDisabled} />)
    expect(wrapper).toMatchSnapshot()
  })
})
