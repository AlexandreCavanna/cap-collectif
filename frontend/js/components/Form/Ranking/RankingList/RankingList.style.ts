import type { StyledComponent } from 'styled-components'
import styled from 'styled-components'
import colors from '~/utils/colors'
import { mediaQueryMobile, bootstrapGrid } from '~/utils/sizes'
import { spaceItemPosition } from '~/components/Ui/DragnDrop/List/List.style'

const RankingListContainer: StyledComponent<any, {}, HTMLDivElement> = styled.div.attrs({
  className: 'rankingList',
})`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;

  .separator {
    margin-top: 90px; /* (height separator + height wrapper-item-container (empty) ) / 2 */
    font-size: 24px;
    color: ${colors.lightGray};
  }

  @media (max-width: ${bootstrapGrid.mdMax}px) {
    .list-dragndrop {
      width: 48%;

      .list__item {
        width: 100%;
      }
    }

    .separator {
      display: none;
    }
  }

  @media (max-width: ${mediaQueryMobile.maxWidth}) {
    .list-dragndrop {
      width: auto;
    }

    .list-dragndrop:first-child {
      padding-left: ${spaceItemPosition}px;
    }
  }
`
export default RankingListContainer
