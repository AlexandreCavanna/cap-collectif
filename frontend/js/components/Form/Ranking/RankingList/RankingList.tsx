import { $Values } from 'utility-types'
import React, { useState, useEffect } from 'react'
import type { DropResult, DragUpdate } from 'react-beautiful-dnd'
import 'react-beautiful-dnd'
import isEqual from 'lodash/isEqual'
import { useIntl } from 'react-intl'
import Context from '~/components/Ui/DragnDrop/Context/Context'
import List from '~/components/Ui/DragnDrop/List/List'
import Item from '~/components/Ui/DragnDrop/Item/Item'
import RankingLabel from '../RankingLabel/RankingLabel'
import RankingListContainer from './RankingList.style'
import usePrevious from '~/utils/hooks/usePrevious'
import { swap, moveItemOnAvailable, moveItem, formatDataDraggable } from '~/utils/dragNdrop'
import config from '~/config'
import type { FieldsProps } from '../Ranking'
import '../Ranking'
import type { Field } from '../../Form.type'
import '../../Form.type'
import VisuallyHidden from '~ds/VisuallyHidden/VisuallyHidden'

const ID_LIST: {
  CHOICES: 'ranking__choices'
  SELECTION: 'ranking__selection'
} = {
  CHOICES: 'ranking__choices',
  SELECTION: 'ranking__selection',
}
type RankingListProps = {
  id: string
  onChange: (arg0: Array<Field>) => void
  dataForm: FieldsProps
  isDisabled?: boolean
}
type PreviewDraggable = {
  draggableSource: {
    draggableId: string
    droppableId: string
  }
  draggableIdDestination: string
} | null
type ListItems = Record<$Values<typeof ID_LIST>, Array<Field>>

const fillRankingWithEmpty = (list, totalList) =>
  totalList - list.length > 0 ? [...list, ...Array(totalList - list.length).fill(null)] : list

const fillOnlyEmpty = length => [...Array(length).fill(null)]

const getTotalChoice = (choice, selection) => {
  let total = 0
  if (choice) total += choice.length
  if (selection) total += selection.length
  return total
}

const RankingList = ({ dataForm, isDisabled, onChange, id }: RankingListProps) => {
  const intl = useIntl()
  const { choices: formChoices, values: formSelection = [] } = dataForm
  const totalChoice: number = getTotalChoice(formChoices, formSelection)
  const [previewDraggable, setPreviewDraggable] = useState<PreviewDraggable>(null)
  const [choices, setChoices] = useState<Array<Field>>(
    formChoices && formChoices.length
      ? fillRankingWithEmpty(formChoices, totalChoice)
      : fillOnlyEmpty(formSelection && formSelection.length),
  )
  // fill of null to get empty draggable item
  const [selection, setSelection] = useState<Array<Field>>(
    formSelection && formSelection.length
      ? fillRankingWithEmpty(formSelection, totalChoice)
      : fillOnlyEmpty(formChoices.length),
  )
  const previousSelection = usePrevious({
    selection,
  })
  useEffect(() => {
    const clearSelection = selection.filter(Boolean)
    const clearChoices = choices.filter(Boolean)
    const clearPreviousSelection = previousSelection && previousSelection.selection.filter(Boolean)

    if (!isEqual(clearSelection, clearPreviousSelection)) {
      return onChange(clearSelection)
    }

    if (!isEqual(clearSelection, formSelection)) {
      setSelection(
        formSelection && formSelection.length
          ? fillRankingWithEmpty(formSelection, totalChoice)
          : [...Array(formChoices.length).fill(null)],
      )
    }

    if (!isEqual(clearChoices, formChoices)) {
      setChoices(
        formChoices && formChoices.length
          ? fillRankingWithEmpty(formChoices, totalChoice)
          : [...Array(formSelection && formSelection.length).fill(null)],
      )
    }
  }, [selection, choices, onChange, formSelection, formChoices, totalChoice, previousSelection])

  const getList = (listId: string) => (listId === ID_LIST.CHOICES ? choices : selection)

  const getDestinationIndex = (list, draggableIdDestination) => {
    let destinationIndex = list.findIndex(item => item && item.id === draggableIdDestination)

    // empty destination
    if (destinationIndex === -1) {
      destinationIndex = Number(draggableIdDestination.match(/\d/g))
    }

    return destinationIndex
  }

  const onSwap = (source, combine) => {
    const { draggableId: draggableIdDestination } = combine
    const currentList = getList(source.droppableId)
    const destinationIndex = getDestinationIndex(currentList, draggableIdDestination)
    const items = swap(currentList, source.index, destinationIndex)
    if (source.droppableId === ID_LIST.SELECTION) setSelection(items)
    if (source.droppableId === ID_LIST.CHOICES) setChoices(items)
  }

  const onDragInOtherList = (source, combine) => {
    const { draggableId: draggableIdDestination, droppableId: droppableIdDestination } = combine
    const destinationList = getList(combine.droppableId)
    const destinationIndex = getDestinationIndex(destinationList, draggableIdDestination)
    const formatedDestination = formatDataDraggable(destinationIndex, droppableIdDestination)
    const listUpdated: ListItems = moveItem(
      getList(source.droppableId),
      getList(droppableIdDestination),
      source,
      formatedDestination,
      dataForm.choices.length,
    )
    setChoices(listUpdated[ID_LIST.CHOICES])
    setSelection(listUpdated[ID_LIST.SELECTION])
  }

  const onDragEnd = (result: DropResult) => {
    const { source, destination, combine } = result
    // dropped outside the list
    if (!destination && !combine) return

    // same list (interchange)
    if (combine && source.droppableId === combine.droppableId) {
      onSwap(source, combine)
    }

    // same list (reorder)
    if (destination && source.droppableId === destination.droppableId) {
      const currentList = getList(source.droppableId)
      const items = swap(currentList, source.index, destination.index)
      if (source.droppableId === ID_LIST.SELECTION) setSelection(items)
      if (source.droppableId === ID_LIST.CHOICES) setChoices(items)
    }

    // from list to another one
    if (combine && source.droppableId !== combine.droppableId) {
      onDragInOtherList(source, combine)
    }

    setPreviewDraggable(null)
  }

  const moveWithoutDrag = (currentIndex, from: $Values<typeof ID_LIST>, to: $Values<typeof ID_LIST>) => {
    const source = formatDataDraggable(currentIndex, from)
    const listUpdated: ListItems = moveItemOnAvailable(getList(from), getList(to), source, to)
    setChoices(listUpdated[ID_LIST.CHOICES])
    setSelection(listUpdated[ID_LIST.SELECTION])
    setPreviewDraggable(null)
  }

  const onDragUpdate = (result: DragUpdate) => {
    const { combine, draggableId, source } = result
    if (!combine) setPreviewDraggable(null)

    if (combine) {
      if (!previewDraggable || previewDraggable.draggableIdDestination !== combine.draggableId) {
        setPreviewDraggable({
          draggableSource: {
            draggableId,
            droppableId: source.droppableId,
          },
          draggableIdDestination: combine.draggableId,
        })
      }
    }
  }

  const dataPreview = previewDraggable
    ? getList(previewDraggable.draggableSource.droppableId).find(
        item => item && item.id === previewDraggable.draggableSource.draggableId,
      )
    : null
  return (
    <RankingListContainer ariaLabelledby={`label-${id}`} tabIndex={0} role="group">
      <VisuallyHidden id="drag-instructions">
        {intl.formatMessage({
          id: 'aria-ranking-item-instructions',
        })}
      </VisuallyHidden>

      <Context isDisabled={isDisabled} onDragEnd={onDragEnd} onDragUpdate={onDragUpdate}>
        <List
          id={ID_LIST.CHOICES}
          title="global.form.ranking.pickBox.title"
          isDisabled={isDisabled}
          isCombineEnabled
          isCombineOnly
        >
          {choices.map((choice, i) =>
            choice && choice.id ? (
              <Item id={choice.id} key={choice.id} position={i} isDisabled={isDisabled}>
                <RankingLabel
                  {...choice}
                  isDisabled={isDisabled}
                  onPick={() => moveWithoutDrag(i, ID_LIST.CHOICES, ID_LIST.SELECTION)}
                />
              </Item>
            ) : (
              <Item
                id={`${ID_LIST.CHOICES}-${i}`}
                key={`${ID_LIST.CHOICES}-${i}`}
                position={i}
                preview={
                  previewDraggable &&
                  dataPreview &&
                  previewDraggable.draggableIdDestination === `${ID_LIST.CHOICES}-${i}` && (
                    <RankingLabel {...dataPreview} />
                  )
                }
              />
            ),
          )}
        </List>

        {!config.isMobile && <i className="separator cap-arrow-2" aria-hidden />}

        <List
          id={ID_LIST.SELECTION}
          title="global.form.ranking.choiceBox.title"
          isDisabled={isDisabled}
          isCombineEnabled
          isCombineOnly
          hasPositionDisplayed
        >
          {selection.map((item, j) =>
            item && item.id ? (
              <Item
                id={item.id}
                key={item.id}
                position={j}
                isDisabled={isDisabled}
                onRemove={() => moveWithoutDrag(j, ID_LIST.SELECTION, ID_LIST.CHOICES)}
              >
                <RankingLabel {...item} isDisabled={isDisabled} isSelected />
              </Item>
            ) : (
              <Item
                id={`${ID_LIST.SELECTION}-${j}`}
                key={`${ID_LIST.SELECTION}-${j}`}
                position={j}
                preview={
                  previewDraggable &&
                  dataPreview &&
                  previewDraggable.draggableIdDestination === `${ID_LIST.SELECTION}-${j}` && (
                    <RankingLabel {...dataPreview} />
                  )
                }
              />
            ),
          )}
        </List>
      </Context>
    </RankingListContainer>
  )
}

export default RankingList
