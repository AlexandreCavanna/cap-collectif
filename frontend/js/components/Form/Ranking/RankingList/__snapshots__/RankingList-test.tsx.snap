// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RankingList /> should render correctly 1`] = `
<RankingListstyle__RankingListContainer
  ariaLabelledby="label-123-rakintlist"
  role="group"
  tabIndex={0}
>
  <VisuallyHidden
    id="drag-instructions"
  >
    aria-ranking-item-instructions
  </VisuallyHidden>
  <Context
    isDisabled={false}
    onDragEnd={[Function]}
    onDragUpdate={[Function]}
  >
    <List
      id="ranking__choices"
      isCombineEnabled={true}
      isCombineOnly={true}
      isDisabled={false}
      title="global.form.ranking.pickBox.title"
    >
      <Item
        id="24"
        isDisabled={false}
        key="24"
        position={0}
      >
        <RankingLabel
          id="24"
          isDisabled={false}
          label="Maxime Arrouard"
          onPick={[Function]}
        />
      </Item>
      <Item
        id="25"
        isDisabled={false}
        key="25"
        position={1}
      >
        <RankingLabel
          id="25"
          isDisabled={false}
          label="Spylou Super Sayen"
          onPick={[Function]}
        />
      </Item>
      <Item
        id="26"
        isDisabled={false}
        key="26"
        position={2}
      >
        <RankingLabel
          id="26"
          isDisabled={false}
          label="Cyril Lage"
          onPick={[Function]}
        />
      </Item>
      <Item
        id="27"
        isDisabled={false}
        key="27"
        position={3}
      >
        <RankingLabel
          id="27"
          isDisabled={false}
          label="Jpec Superman"
          onPick={[Function]}
        />
      </Item>
    </List>
    <i
      aria-hidden={true}
      className="separator cap-arrow-2"
    />
    <List
      hasPositionDisplayed={true}
      id="ranking__selection"
      isCombineEnabled={true}
      isCombineOnly={true}
      isDisabled={false}
      title="global.form.ranking.choiceBox.title"
    >
      <Item
        id="ranking__selection-0"
        key="ranking__selection-0"
        position={0}
        preview={null}
      />
      <Item
        id="ranking__selection-1"
        key="ranking__selection-1"
        position={1}
        preview={null}
      />
      <Item
        id="ranking__selection-2"
        key="ranking__selection-2"
        position={2}
        preview={null}
      />
      <Item
        id="ranking__selection-3"
        key="ranking__selection-3"
        position={3}
        preview={null}
      />
    </List>
  </Context>
</RankingListstyle__RankingListContainer>
`;

exports[`<RankingList /> should render correctly when disabled 1`] = `
<RankingListstyle__RankingListContainer
  ariaLabelledby="label-123-rakintlist"
  role="group"
  tabIndex={0}
>
  <VisuallyHidden
    id="drag-instructions"
  >
    aria-ranking-item-instructions
  </VisuallyHidden>
  <Context
    isDisabled={true}
    onDragEnd={[Function]}
    onDragUpdate={[Function]}
  >
    <List
      id="ranking__choices"
      isCombineEnabled={true}
      isCombineOnly={true}
      isDisabled={true}
      title="global.form.ranking.pickBox.title"
    >
      <Item
        id="24"
        isDisabled={true}
        key="24"
        position={0}
      >
        <RankingLabel
          id="24"
          isDisabled={true}
          label="Maxime Arrouard"
          onPick={[Function]}
        />
      </Item>
      <Item
        id="25"
        isDisabled={true}
        key="25"
        position={1}
      >
        <RankingLabel
          id="25"
          isDisabled={true}
          label="Spylou Super Sayen"
          onPick={[Function]}
        />
      </Item>
      <Item
        id="26"
        isDisabled={true}
        key="26"
        position={2}
      >
        <RankingLabel
          id="26"
          isDisabled={true}
          label="Cyril Lage"
          onPick={[Function]}
        />
      </Item>
      <Item
        id="27"
        isDisabled={true}
        key="27"
        position={3}
      >
        <RankingLabel
          id="27"
          isDisabled={true}
          label="Jpec Superman"
          onPick={[Function]}
        />
      </Item>
    </List>
    <i
      aria-hidden={true}
      className="separator cap-arrow-2"
    />
    <List
      hasPositionDisplayed={true}
      id="ranking__selection"
      isCombineEnabled={true}
      isCombineOnly={true}
      isDisabled={true}
      title="global.form.ranking.choiceBox.title"
    >
      <Item
        id="ranking__selection-0"
        key="ranking__selection-0"
        position={0}
        preview={null}
      />
      <Item
        id="ranking__selection-1"
        key="ranking__selection-1"
        position={1}
        preview={null}
      />
      <Item
        id="ranking__selection-2"
        key="ranking__selection-2"
        position={2}
        preview={null}
      />
      <Item
        id="ranking__selection-3"
        key="ranking__selection-3"
        position={3}
        preview={null}
      />
    </List>
  </Context>
</RankingListstyle__RankingListContainer>
`;
